jane/doc/local-reference.md

```diff
@@ -442,39 +442,81 @@ be the caller's (or that of the caller's caller, etc., if the caller is also
 
-In the previous section, a function can return local values by not having its own
-region; as a result, it works in the caller's region. This has downsides;
-consider the following example:
 ```
 let f (local_ x) = local_
   let local_ y = (complex computation on x) in
   if y then local_ None
   else local_ (Some x)
 ```
-function `f` will allocate in caller's region to store
-intermediate/temporary data for the complex computation, and the allocation
-will stay in the region even after `f` returns. They would be released
-only when the program exits the caller's region. We hope that the temporary
-allocation are released upon `f` returning and rewrite the example as
+
+The function `f` allocates memory within the caller's region to store
+intermediate and temporary data for the complex computation. This allocation
+remains in the region even after `f` returns and is released only when the
+program exits the caller's region. To allow temporary allocations to be released
+upon the function's return, we can rewrite the example as follows:
+
+
 ```
 let f (local_ x) =
   let local_ y = (complex computation on x) in
   if y then [%exclave] (local_ None)
   else [%exclave] (local_ (Some x))
 ```
-The new primitive `[%exclave]` early-terminates the current region, and runs
-whatever follows in the outer region. In this example, the function `f` still
-has its own region, where the allocation of the complex computation will happen.
-This region will be early-terminated by `[%exclave]`, releasing all the
-temporary allocations. Next, both `local_ None` and `local_ (Some x)` are "local"
-relative to the outer region and thus allowed to escape. In summary, we have
-temporary allocations on stack and released promptly, and have the result
-allocations on stack and escaped.
-
-`[%exclave]` terminates the current region, and all values therein are
-deleted. For example, the following is error because `x` is deleted after
-[%exclave] and cannot be refered to.
+
+The new primitive `[%exclave]` terminates the current region early and executes
+the subsequent code in the outer region. In this example, the function `f`
+retains its own region where the allocation for the complex computation occurs.
+This region is terminated by `[%exclave]`, releasing all temporary allocations.
+Both `local_ None` and `local_ (Some x)` are considered "local" relative to the
+outer region and are allowed to escape. In summary, we have temporary
+allocations on the stack that are promptly released and result allocations on
+the stack that can escape.
+
+
+Here is another example in which the stack usage can be improved asymptotically
+by applying `[%exclave]`:
+
+
+```
+let rec maybe_length p l = local_
+  match l with
+  | [] -> Ok { res = 0 }
+  | x :: xs ->
+      match p x with
+      | false -> None
+      | true ->
+          match count_result f xs with
+          | None as r -> r
+          | Some { res = count } ->
+              Some { res = count + 1 }
+```
```

Comment:
Let's just take this example out for now. I'll add a better one in a later PR.

Reply:
is it better to have a silly but working example, than to have no example?